### DISCUSSION TITLE
Pitchbend: Force "middle position" to 63/63 ?

### DISCUSSION THREAD

**Discussion by jonvog** [tttapa/Control-Surface#478] [Q&A]
I was wondering if there is a possibility to force the output of PBPotentiometer to the real middle point of 63/63 when it is in it's default position (more or less the analog middle). Like an extra broad hysteresis range in the middle or something like that? I'm not sure if I'm making myself clear... But it would be great, so when performing pitch bends, the tuning would go back to a perfect zero and not stay a little bit under or over 0.

**Comment by tttapa** [comment]
You should be able to do this in the mapping function. Instead of having a single linear mapping function, make it piecewise linear. It's easiest if you just draw it: the _x_ axis representing the raw input (the range of your potentiometer), and the _y_ axis representing the output of the mapping function (in the range [0, `FilteredAnalog<>::getMaxRawValue()`]).

```
  |      /
  |   _ /
y | /
  |/______
      x
```
You can program this using two `if` statements and the `map()` function.

Also see the discussion in #214.

Note that in that discussion, the offset of 64 is determined as follows:  
&emsp;&emsp;2<sup>B</sup> / 2<sup>R</sup> / 2,  
where B is the number of bits of the raw signal (14 by default, but different if you change the analog filter settings), and R is the resolution of the MIDI output (7 in the case of control change, 10 in the case of pitch bend). If I'm not mistaken, that should work out to
```cpp
constexpr analog_t offset = (FilteredAnalog<>::getMaxRawValue() + 1) >> (10 + 1);
```

**Reply by jonvog** [reply]
Great, generally this works for me. I didn't understand how to implement the offset though. Now I have an offset of only -9 at the "0" position (in 14bit PB), that works for me. I also tried adding or subtracting a value like `else { return maxRawValue / 2 + 1; }` but I only could get the output to -9 or 8, nothing in between. 

Honestly I couldn't figure out, how to solve this, but it doesn't really matter in real life for me, so I don't care too much.

This is what I ended up with (I'm sure there is a cleaner way, but that's what I could figure out):

```
// Mapping
constexpr unsigned long maxRawValue = FilteredAnalog<>::getMaxRawValue();
constexpr analog_t minimumValue = 20400 * maxRawValue / 65535; // The filtered value read when potentiometer is at the 0% position
constexpr analog_t maximumValue = 46000 * maxRawValue / 65535; // The filtered value read when potentiometer is at the 100% position
constexpr analog_t margin = 150;
constexpr analog_t middlePoint = minimumValue + (maximumValue - minimumValue) / 2;

analog_t mappingFunction(analog_t raw) {
  // make sure that the analog value is between the minimum and maximum
  raw = constrain(raw, minimumValue, maximumValue);
  
  // map the value from [minimumValue, maximumValue] to [0, maxRawValue]
  if (raw <= middlePoint - margin){
   return map(raw, minimumValue, middlePoint, 0, maxRawValue / 2);
  }
  else if (raw >= middlePoint + margin){
   return map(raw, middlePoint, maximumValue, maxRawValue / 2, maxRawValue);
  }
  else {
   return maxRawValue / 2;
  }
}
```

Thanks again for your help!

**Comment by josehdx** [comment]
Hello, I need some help to force this CC63 value as well, when i return to center. i can not even understand this equation:   2B / 2R / 2,
 i mean, The B for me is 7 as i modified the resolution to avoid noise on the class:  (MIDIFilteredAnalog<PitchBendSender<7>>)

Then the R should be 10 as i am using a pitche bender. but the result is very small= .0625

I assume i am doing the math worng.

The current CC number i got when i return to center is  CC57 or 7353 instead of CC63 or 8192

I also dont know where to instert the mapping code mentioned above, i assume it should be on the class, but better to ask

Thanks

**Comment by josehdx** [comment]
My code

`#include <Control_Surface.h>

#include <AH/Hardware/MultiPurposeButton.hpp>

BluetoothMIDI_Interface midi;

// Instantiate a PBPotentiometer object
PBPotentiometer potentiometer {
  A12,        // Analog pin connected to potentiometer
  CHANNEL_1, // MIDI Channel 1
};

MultiPurposeButton btn{
  
  {0}

};

MultiPurposeButton btn2{
  
  {35}

};


void setup() {
  Control_Surface.begin();
  btn.setLongPressDelay(550);
  btn.setMultiPressDelay(400);
  btn.begin();


  btn2.setLongPressDelay(550);
  btn2.setMultiPressDelay(400);
  btn2.begin();
}

void loop() {


  switch (btn.update()) {
    case btn.None: break;
    case btn.PressStart: break;
    
    case btn.ShortPressRelease: Control_Surface.sendControlChange({MIDI_CC::General_Purpose_Controller_1, CHANNEL_1}, 0);
                                Control_Surface.sendControlChange({MIDI_CC::General_Purpose_Controller_1, CHANNEL_1}, 127);
            
          break;
          
   case btn.LongPress: Control_Surface.sendControlChange({MIDI_CC::General_Purpose_Controller_1, CHANNEL_2}, 0);
                        Control_Surface.sendControlChange({MIDI_CC::General_Purpose_Controller_1, CHANNEL_2}, 127);

       
          break;
     
    
  }

 switch (btn2.update()) {
    case btn2.None: break;
    case btn2.PressStart: break;

    case btn2.ShortPressRelease: Control_Surface.sendControlChange({MIDI_CC::General_Purpose_Controller_1, CHANNEL_3}, 0);
                                Control_Surface.sendControlChange({MIDI_CC::General_Purpose_Controller_1, CHANNEL_3}, 127);
                
          break;

         
   case btn2.LongPress: Control_Surface.sendControlChange({MIDI_CC::General_Purpose_Controller_1, CHANNEL_4}, 0);
                        Control_Surface.sendControlChange({MIDI_CC::General_Purpose_Controller_1, CHANNEL_4}, 127);
    
          break;
 }

  Control_Surface.loop();

}`

**Comment by jonvog** [comment]
TBH unfortunately I don't know enough about all this stuff to help you with your issue and I can't really recall what I did to get it working. But this is my sketch and it works, so maybe you can make sense of it. Just for context: I have 8 potentiometers and 1 PB wheel connected. + I connected a midi keybed to the serial in and forward everything from that keybed to my computer via pipes in the `void setup()` section.

```
#include <Control_Surface.h> // Include the Control Surface library
 
// Instantiate a MIDI over USB interface.
USBMIDI_Interface midiusb;
HardwareSerialMIDI_Interface midiser = Serial1;

// Instantiate MIDI Pipes
MIDI_PipeFactory <1> pipes;
 
// Instantiate an array of CCPotentiometer objects
CCPotentiometer potentiometers[] = {
  {A0, 0x16}, // pot 1 (controller number as hex) - CC22
  {A1, 0x17}, // pot 2 - CC23   
  {A2, 0x18}, // pot 3 - CC24
  {A3, 0x19}, // pot 4 - CC25
  {A4, 0x01}, // fader 1 - CC1
  {A5, 0x0B}, // fader 2 - CC11
  {A6, 0x15}, // fader 3 - CC21
  {A7, 0x07}, // fader 4 - CC7
};

// Instantiate the pitchbend potentiometer
PBPotentiometer pitchBend = {
  A8,        // Analog pin connected to potentiometer
  CHANNEL_1, // MIDI Channel 1 
};

// Mapping Pitchbend
constexpr unsigned long maxRawValue = FilteredAnalog<>::getMaxRawValue();
constexpr analog_t minimumValue = 20400 * maxRawValue / 65535; // The filtered value read when potentiometer is at the 0% position
constexpr analog_t maximumValue = 46000 * maxRawValue / 65535; // The filtered value read when potentiometer is at the 100% position
constexpr analog_t margin = 150;
constexpr analog_t middlePoint = minimumValue + (maximumValue - minimumValue) / 2;

analog_t mappingFunction(analog_t raw) {
  // make sure that the analog value is between the minimum and maximum
  raw = constrain(raw, minimumValue, maximumValue);
  
  // map the value from [minimumValue, maximumValue] to [0, maxRawValue]
  if (raw <= middlePoint - margin){
   return map(raw, minimumValue, middlePoint, 0, maxRawValue / 2);
  }
  else if (raw >= middlePoint + margin){
   return map(raw, middlePoint, maximumValue, maxRawValue / 2, maxRawValue);
  }
  else {
   return maxRawValue / 2;
  }
}

void setup() {
  pitchBend.map(mappingFunction); // Add the mapping function to the potentiometer
  midiusb.setAsDefault(); // (must be done before initializing Control Surface)
  midiser >> pipes >> midiusb; // all incoming midi from Serial is sent to USB
  Control_Surface.begin(); // Initialize Control Surface
  // Serial.begin(115200); // uncomment for using the serial monitor
}
 
void loop() {
  Control_Surface.loop(); // Update the Control Surface
  // Serial.println(pitchBend.getRawValue()); // print the raw PB value to the serial monitor
}
```

**Reply by josehdx** [reply]
Thank you. I will try to adapt it to my case today
